### PR TITLE
When AssignmentSet contains a Ref, bind the real value before mapping to Assignments

### PR DESCRIPTION
**Pull request by jutley**

### CODE REVIEW

**Review by benkehoe — COMMENTED** [review]
Let me think about the implementation. I'm not sure whether I'd rather have a "preprocess" step that performs this, or a callback function that allows resolving references in various places. I would also not want it to dereference single-valued parameters if not necessary.

I'm also planning a fairly substantial revision to the assignment generation, so I want to think about how it fits into that as well. Give me a week.

**Comment by benbridts** [comment]
> I would also not want it to dereference single-valued parameters if not necessary.

I think this fits with the following remark of @jutley 

> Alternatively, the logic of replacing parameter refs with the parameter values can be done in an independent Transform. I'm not aware of any existing solutions for this.

I'm not sure what the consequences would be of nesting transforms though

**Comment by jutley** [comment]
> I'm not sure what the consequences would be of nesting transforms though

The documentation states that the order of macro evaluation is done by evaluating the inner-most macros first, then working outward.
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-macros.html#template-macros-order

**Comment by jutley** [comment]
@benkehoe Any update on this topic, or on the changes to assignment generation?

**Comment by benkehoe** [comment]
Sorry about the delay! I've been overwhelmed with other things recently. I think there are two ways we could go with this. In general, I'd like not to de-reference parameters when we don't have to, so that the processed template retains as much of the semantics of the original as possible.
* If we want to unroll list parameters for any `AssignmentGroup` id parameter (that is, if the OU id is a single value that is a `!Ref` of a `CommaDelimitedList` parameter, it will result in a list of OU ids, and if the OU id is a list, and an item is a `!Ref` of a `CommandDelimitedList` paramter, it will result in those OU ids be expanded within the list, and same for principal id, permission set ARN, etc.), then we should add a `reference_resolver` parameter to `Config.load_resource_properties()` that is a callable.
* If we don't care about unrolling lists (we assume every reference is , then I *think* we only need this for OU ids, right? In which case I think we would expand the [ou_fetcher](https://github.com/benkehoe/aws-sso-util/blob/c69c70ecf41eabcad251a489df01833fc35cae76/cli/src/aws_sso_util/cfn_lib/macro.py#L105) with the capability to resolve references.
What do you think?

**Comment by jutley** [comment]
> I'd like not to de-reference parameters when we don't have to

I am a bit mixed on this point, but I think I agree. Trying to determine when to de-reference a parameter probably adds some complexity, but I am realizing that the macro is going to have to treat single-value parameters and list parameters differently anyway, so maybe the additional complexity is trivial.

My only remaining concern here is that a user may look at the processed template and be confused as to why some parameters were de-referenced and some were not.

> and an item is a !Ref of a CommandDelimitedList paramter, it will result in those OU ids be expanded within the list

I think I actually disagree on this part! The existing model seems to let us have separate target entries to avoid having to embed lists within lists. For example, instead of writing this:

```
Target:
  - Type: AWS_ACCOUNT
    Id:
      - 000000000000
      - 111111111111
      - !Ref CommaSeparatedAccounts1
      - !Ref CommaSeparatedAccounts2
```

We can instead write this:

```
Target:
  - Type: AWS_ACCOUNT
    Id:
      - 000000000000
      - 111111111111
  - Type: AWS_ACCOUNT
    Id: !Ref CommaSeparatedAccounts1
  - Type: AWS_ACCOUNT
    Id: !Ref CommaSeparatedAccounts2
```

This way we can reduce the number of scenarios that need to be accounted for in the de-reference logic.

> If we don't care about unrolling lists

Personally, I think it is worth unrolling the list parameters. The accounts that organizations want to target will not always land cleanly within an OU, and I can imagine that some may want to have lists of OUs.

---

So to summarize, I think de-referencing should happen when 1) the parameter is a list, or 2) the parameter is for an OU (both may be true). We may want to de-reference in all cases if it leads to a better user experience.
